### PR TITLE
Add Azure Named Key Credential in Core

### LINKED ISSUE #19881: Azure.Core: AzureNamedKeyCredential
# Summary

When using a shared key for authorization, some services require that the name of the key be used as a component of the algorithm to form the authorization token.  For these cases, `AzureKeyCredential` and `AzureSasCredential` aren't adequate as they track only a single value.  A new credential type, `AzureNamedKeyCredential` has been approved to address this gap.

# Goals

- Offer a token credential suitable for working with any service needing a named key that follows patterns established by the other credential types offered in `Azure.Core` and `Azure.Identity`.

- Developers should be able to construct the credential by providing a string-based name and key.

- Developers should be able to update the name and key after construction to support rolling credentials or using a new access policy.  Updating credential information should be done in-place and not require creating a new credential instance.

- Developers should be able to read the name of they key, in order to support caching within an application so that it can be easily identified when updates are needed.

- The Azure SDKs should be able to read the name and key, ensuring they reflect the most current values.

- Support updates and reads concurrently with consistent behavior.  Developers should be able to update the credential at any time without concern for whether the SDK is reading from it.

# Non-Goals

- Providing a friendly and discoverable means for developers to read the key from the credential; it is preferable to hide the key or discourage use.

- Allowing the name or key to be updated independently; both elements will be required when updating a credential.

# Success Criteria

- An `AzureNamedKeyCredential` type has been implemented, following language guidelines and idioms.

- The tests necessary for its validation have been created and pass reliably.

- The existing test suite continues to produce deterministic results and pass reliably.

# Related

- [.NET AzureNamedKeyCredential Design Options](https://gist.github.com/jsquire/861fb365fa672dd62530500ad2bc0db9)

### PR DESCRIPTION
**Pull request by g2vinay**
Fixes #19881 

Ports the arch-approved design for AzureNamedKeyCredential. in .NET to Java.

### CODE REVIEW

File: sdk/core/azure-core/src/main/java/com/azure/core/credential/AzureNamedKeyCredential.java [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+            throw logger.logExceptionAsError(new IllegalArgumentException("'name' cannot be empty."));
+        }
+        if (key.isEmpty()) {
+            throw logger.logExceptionAsError(new IllegalArgumentException("'key' cannot be empty."));
+        }
+        this.credentials = new Tuple<>(name, key);
+        return this;
+    }
+
+    private static class Tuple<X, Y> {
```
Comment (alzimmermsft) [inline sdk/core/azure-core/src/main/java/com/azure/core/credential/AzureNamedKeyCredential.java:78]: Let's just simplify this to be non-generic and having the field names being what they represent (`name`, `key`)

File: sdk/core/azure-core/src/main/java/com/azure/core/credential/AzureNamedKeyCredential.java [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+    public String getName() {
+        return credentials.getLeft();
+    }
+
+    /**
+     * Rotates the {@code name} and  {@code key} associated to this credential.
+     *
+     * @param name The name of the key credential.
+     * @param key The new key to associated with this credential.
+     * @return The updated {@code ApiKeyCredential} object.
```
Comment (alzimmermsft) [inline sdk/core/azure-core/src/main/java/com/azure/core/credential/AzureNamedKeyCredential.java:61]: nit: Wrong name

```suggestion
     * @return The updated {@code AzureNamedKeyCredential} object.
```

File: sdk/core/azure-core/src/test/java/com/azure/core/credential/CredentialsTests.java [hunk 3]
```diff
@@ -13,16 +13,21 @@
 import com.azure.core.http.policy.HttpPipelinePolicy;
 import org.junit.jupiter.api.Assertions;
 import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtensionContext;
 import org.junit.jupiter.params.ParameterizedTest;
-import org.junit.jupiter.params.provider.CsvSource;
+import org.junit.jupiter.params.provider.*;
```
Comment (alzimmermsft) [inline sdk/core/azure-core/src/test/java/com/azure/core/credential/CredentialsTests.java:18]: nit: `*` import

File: sdk/core/azure-core/src/main/java/com/azure/core/credential/AzureNamedKeyCredential.java (hunk 1 above)
Comment (g2vinay) [reply sdk/core/azure-core/src/main/java/com/azure/core/credential/AzureNamedKeyCredential.java:78]: Kept it generic in the hope to extract it as a Util offering in azure-core.

Comment (g2vinay) [reply sdk/core/azure-core/src/main/java/com/azure/core/credential/AzureNamedKeyCredential.java]: changed it non generic for now.
Might be good o extract it out as Util. I remember KV uses something similar to this as impl detail.

File: sdk/core/azure-core/src/main/java/com/azure/core/credential/AzureNamedKeyCredential.java [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+    }
+
+    /**
+     * Retrieves the name associated to this credential.
+     *
+     * @return The key being used to authorize requests.
+     */
+    public String getName() {
+        return credentials.getName();
+    }
```
Comment (srnagar) [inline sdk/core/azure-core/src/main/java/com/azure/core/credential/AzureNamedKeyCredential.java:54]: We shouldn't have separate getter methods for name and key. We should have one getter method that returns `AzureNamedKey` that contains both fields.

File: sdk/core/azure-core/src/main/java/com/azure/core/credential/AzureNamedKeyCredential.java [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+            throw logger.logExceptionAsError(new IllegalArgumentException("'name' cannot be empty."));
+        }
+        if (key.isEmpty()) {
+            throw logger.logExceptionAsError(new IllegalArgumentException("'key' cannot be empty."));
+        }
+        this.credentials = new CredentialTuple(name, key);
+        return this;
+    }
+
+    private static class CredentialTuple {
```
Comment (srnagar) [inline sdk/core/azure-core/src/main/java/com/azure/core/credential/AzureNamedKeyCredential.java:78]: This should be a public class and should be named `AzureNamedKey`.

File: sdk/core/azure-core/src/main/java/com/azure/core/credential/AzureNamedKeyCredential.java (hunk 4 above)
Comment (g2vinay) [reply sdk/core/azure-core/src/main/java/com/azure/core/credential/AzureNamedKeyCredential.java]: updated

File: sdk/core/azure-core/src/main/java/com/azure/core/credential/AzureNamedKeyCredential.java (hunk 5 above)
Comment (g2vinay) [reply sdk/core/azure-core/src/main/java/com/azure/core/credential/AzureNamedKeyCredential.java]: updated.
This change won't be consistent with .NET though.
But, its probably fine to be idiomatically different.

File: sdk/core/azure-core/src/main/java/com/azure/core/credential/AzureNamedKey.java [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+// Licensed under the MIT License.
+
+package com.azure.core.credential;
+
+/**
+ * Represents a credential bag containing the key and the name of the key.
+ *
+ * @see AzureNamedKeyCredential
+ */
+public class AzureNamedKey {
```
Comment (srnagar) [inline sdk/core/azure-core/src/main/java/com/azure/core/credential/AzureNamedKey.java:11]: Make this `final` and add `@Immutable` annotation.

File: sdk/core/azure-core/src/main/java/com/azure/core/credential/AzureNamedKeyCredential.java [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+    public AzureNamedKeyCredential update(String name, String key) {
+        Objects.requireNonNull(name, "'name' cannot be null.");
+        Objects.requireNonNull(key, "'key' cannot be null.");
+        if (name.isEmpty()) {
+            throw logger.logExceptionAsError(new IllegalArgumentException("'name' cannot be empty."));
+        }
+        if (key.isEmpty()) {
+            throw logger.logExceptionAsError(new IllegalArgumentException("'key' cannot be empty."));
+        }
+        this.credentials = new AzureNamedKey(name, key);
```
Comment (srnagar) [inline sdk/core/azure-core/src/main/java/com/azure/core/credential/AzureNamedKeyCredential.java:65]: This is a duplicate of the code in the constructor. Extract this into a method.

File: sdk/core/azure-core/src/main/java/com/azure/core/credential/AzureNamedKey.java [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+    private final String name;
+    private final String key;
+
+    AzureNamedKey(String name, String key) {
```
Comment (g2vinay) [inline sdk/core/azure-core/src/main/java/com/azure/core/credential/AzureNamedKey.java:15]: didn't make this public ,as its purpose is to extract the credential details currently from AzureNamedKeyCredential.

File: sdk/core/azure-core/src/main/java/com/azure/core/credential/AzureNamedKeyCredential.java [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+     */
+    public AzureNamedKey getAzureNamedKey() {
+        return this.credentials;
+    }
+
+    /**
+     * Rotates the {@code name} and  {@code key} associated to this credential.
+     *
+     * @param name The name of the key credential.
+     * @param key The new key to associated with this credential.
```
Comment (srnagar) [inline sdk/core/azure-core/src/main/java/com/azure/core/credential/AzureNamedKeyCredential.java:51]:
```suggestion
     * @param key The new key to be associated with this credential.
```

File: sdk/core/azure-core/src/main/java/com/azure/core/credential/AzureNamedKeyCredential.java [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+     * @return The {@link AzureNamedKey} containing the name and key .
+     */
+    public AzureNamedKey getAzureNamedKey() {
+        return this.credentials;
+    }
+
+    /**
+     * Rotates the {@code name} and  {@code key} associated to this credential.
+     *
+     * @param name The name of the key credential.
```
Comment (srnagar) [inline sdk/core/azure-core/src/main/java/com/azure/core/credential/AzureNamedKeyCredential.java:50]:
```suggestion
     * @param name The new name of the key credential.
```

File: sdk/core/azure-core/src/main/java/com/azure/core/credential/AzureNamedKey.java [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+     * @return The key being used to authorize requests.
+     */
+    public String getKey() {
+        return key;
+    }
+
+    /**
+     * Retrieves the name associated to this credential.
+     *
+     * @return The key being used to authorize requests.
```
Comment (srnagar) [inline sdk/core/azure-core/src/main/java/com/azure/core/credential/AzureNamedKey.java]:
```suggestion
     * @return The name of the key used to authorize requests.
```

**Review by srnagar — APPROVED** [review]

File: sdk/core/azure-core/src/main/java/com/azure/core/credential/AzureNamedKeyCredential.java [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+        validateInputParameters(name, key);
+        this.credentials = new AzureNamedKey(name, key);
+    }
+
+    /**
+     * Retrieves the {@link AzureNamedKey} containing the name and key associated with this credential.
+     *
+     * @return The {@link AzureNamedKey} containing the name and key .
+     */
+    public AzureNamedKey getAzureNamedKey() {
```
Comment (alzimmermsft) [inline sdk/core/azure-core/src/main/java/com/azure/core/credential/AzureNamedKeyCredential.java:36]: This is returning non-public API in a public API

File: sdk/core/azure-core/src/main/java/com/azure/core/credential/AzureNamedKeyCredential.java [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+
+    private void validateInputParameters(String name, String key) {
+        Objects.requireNonNull(name, "'name' cannot be null.");
+        Objects.requireNonNull(key, "'key' cannot be null.");
+        if (name.isEmpty()) {
+            throw logger.logExceptionAsError(new IllegalArgumentException("'name' cannot be empty."));
+        }
+        if (key.isEmpty()) {
+            throw logger.logExceptionAsError(new IllegalArgumentException("'key' cannot be empty."));
+        }
```
Comment (alzimmermsft) [inline sdk/core/azure-core/src/main/java/com/azure/core/credential/AzureNamedKeyCredential.java:63]: Should this validation be moved in `AzureNamedKey`?

File: sdk/core/azure-core/src/main/java/com/azure/core/credential/AzureNamedKeyCredential.java (hunk 12 above)
Comment (g2vinay) [reply sdk/core/azure-core/src/main/java/com/azure/core/credential/AzureNamedKeyCredential.java:45]: AzureNamedKey is a public class.

File: sdk/core/azure-core/src/main/java/com/azure/core/credential/AzureNamedKeyCredential.java (hunk 13 above)
Comment (g2vinay) [reply sdk/core/azure-core/src/main/java/com/azure/core/credential/AzureNamedKeyCredential.java:72]: we can refactor it later, if and when we make its constructor public.

**Review by alzimmermsft — APPROVED** [review]

**Pull request merged**